src/polyswarmd/websockets/json_schema.py

```diff
@@ -1,88 +1,122 @@
+
+        srckey: str - Get the field with the same name from `instance`
+        srckey: callable - Run this function with `instance` as an argument
+
+    If srckey is not present at all, it attempts to fetch the value from `source`
+    with the same name/key as the definition.
+
+    >>> make_range = lambda key, src: range(1, src[key])
+    >>> schema = PSJSONSchema({
+    ... 'properties': {
+    ...     'a': {'type':'string'},
+    ...     'b': {'type':'string', 'srckey': 'b_src'},
+    ...     'c': {'type':'string', 'srckey': 'c_src'},
+    ...     'x': {'type':'integer'},
+    ...     'range': {'type': 'array', 'items': 'string', 'srckey': make_range},
+    ...     'xs': {'type': 'array', 'items': 'integer' }}})
+    >>> instance = { 'a': "1", 'b_src': "2", 'c_src': 3, 'x': 4, 'xs': ["5","6"], 'range': 3 }
+    >>> schema.extract(instance)
+    {'a': '1', 'b': '2', 'c': '3', 'x': 4, 'range': ['1', '2'], 'xs': [5, 6]}
     """
-    result = {}
-    for key, pschema in schema['properties'].items():
-        value = None
-        if '$#from' in pschema:
-            value = source[pschema['$#from']]
-        elif '$#fetch' in pschema:
-            value = pschema['$#fetch'](source, key, pschema)
-        else:
-            try:
-                value = source[key]
-            except (AttributeError, IndexError, KeyError) as e:
-                raise KeyError('Schema key not found in source ', key, source, schema)
-
-        if pschema.get('$#convert'):
-            result[key] = _apply_conversion(_apply_format(value, pschema), pschema)
-        else:
-            result[key] = value
-
-    return result
-
-
-_formatters = {'uuid': lambda x: uuid.UUID(int=x)}
-
-
-def _apply_format(value: Any, schema: JSONSchema) -> Any:
-    if 'format' in schema and schema['format'] in _formatters:
-        return _formatters[schema['format']](value)
-    return value
-
-
-_conversions = {
-    'string': str,
-    'integer': int,
-    'number': float,
-    'bool': bool,
-    'boolean': bool,
-}
-
-
-def _apply_conversion(value: Any, schema: JSONSchema) -> Any:
-    dtype = schema.get('type')
-    itype = schema.get('items')
-    if dtype == 'array' and itype in _conversions:
-        # this should really be recursive, but we're not using nested item definitions yet.
-        return [_conversions[itype](v) for v in value]
-    if dtype in _conversions:
-        return _conversions[schema['type']](value)
-    return value
-
-
-if __name__ == "__main__":
-    import doctest
-    doctest.testmod()
+    schema: JSONSchema
+    _extractor: Dict[str, Callable[[Any], Any]]
+
+    _TYPES = {'string': str, 'integer': int, 'number': float, 'boolean': bool, 'array': list}
+
+    _FORMATTERS = {'uuid': lambda x: uuid.UUID(int=x)}
+
+    def __init__(self, schema: Dict[str, Any]):
+        self.schema = cast(JSONSchema, schema)
+        self._extractor = self.build_extractor()
+
+    def visitor(self):
```

Comment:
This was refactored to use the visitor pattern?

Reply:
It previously was, but now it's basically just a way to fetch an iterable of properties.